pkg/sql/schemachanger/eav/tree.go

```diff
@@ -0,0 +1,285 @@
+// Copyright 2021 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package eav
+
+import (
+	"math"
+	"sync"
+	"unsafe"
+
+	"github.com/cockroachdb/cockroach/pkg/util/iterutil"
+	"github.com/cockroachdb/errors"
+	"github.com/google/btree"
+)
+
+// Tree is a data structure for indexing elements.
+type Tree struct {
+	t      *btree.BTree
+	schema Schema
+
+	// dims stores the specifications of all of the indexes in which
+	// all of the elements in the tree are stored. When an entity is
+	// inserted, it is inserted into each of the indexes. The first
+	// entry in the list is the "primary index" which compares entities
+	// based on all attributes.
+	dims []indexSpec
+}
+
+// Schema returns the schema associated with the tree.
+func (t *Tree) Schema() Schema {
+	return t.schema
+}
+
+// NewTree constructs a new Tree with the specified indexes.
+// Note that the schema must not contain more than 64 attributes.
+func NewTree(sc Schema, indexes [][]Attribute) *Tree {
+	t := &Tree{
+		schema: sc,
+		t:      btree.New(8 /* degree */),
```

Comment:
What's the point of having one big b-tree for all indexes instead of one per index? Especially if in the ordering the index takes precedence over the rest of the item. I don't see any benefit for traversals, and insertions are more painful.